tls/s2n_x509_validator.c

```diff
@@ -549,3 +550,64 @@ s2n_cert_validation_code s2n_x509_validator_validate_cert_stapled_ocsp_response(
     return ret_val;
 #endif /* S2N_OCSP_STAPLING_SUPPORTED */
 }
+
+DEFINE_POINTER_CLEANUP_FUNC(STACK_OF(X509)*, wipe_cert_chain);
+
+s2n_cert_validation_code s2n_x509_validator_validate_certificate_signatures(struct s2n_x509_validator *validator,  struct s2n_connection *conn)
+{
+    if (validator->skip_cert_validation || !conn->config->security_policy->certificate_signature_preferences) {
+        return S2N_CERT_OK;
+    }
+    S2N_ERROR_IF((validator->state != VALIDATED && validator->state != OCSP_VALIDATED), S2N_ERR_INVALID_STATE);
+    
+    DEFER_CLEANUP(STACK_OF(X509) *validated_chain = X509_STORE_CTX_get1_chain(validator->store_ctx), wipe_cert_chain_pointer);
+
+    const int certs_in_chain = sk_X509_num(validated_chain);
+
+    /* Do not validate the root certificate */
+    int certs_to_validate = certs_in_chain - 1;
+    
+    for (int i = 0; i < certs_to_validate; i++) {
+        bool out = 0;
+        X509 *cert = sk_X509_value(validated_chain, i);
+
+        GUARD_AS_POSIX(s2n_is_certificate_sig_scheme_supported(conn, cert, &out));
+        if(out == false) {
+            return S2N_CERT_ERR_UNTRUSTED;
+        }
+    }
+
+    return S2N_CERT_OK;
+}
+
+S2N_RESULT s2n_is_certificate_sig_scheme_supported(struct s2n_connection *conn, X509 *x509_cert, bool *out)
+{
+    ENSURE_REF(conn);
+    ENSURE_REF(x509_cert);
+    ENSURE_REF(out);
+
+    int nid = 0;
+
+    #if !defined(LIBRESSL_VERSION_NUMBER)
+        nid = X509_get_signature_nid(x509_cert);
+    #else
+        nid = OBJ_obj2nid(x509_cert->sig_alg->algorithm);
+    #endif
+
+    for (int i = 0; i < conn->config->security_policy->certificate_signature_preferences->count; i++) {
+
+        if (conn->config->security_policy->certificate_signature_preferences->signature_schemes[i]->libcrypto_nid == nid) {
+            /* SHA-1 algorithms are not supported in certificate signatures in TLS1.3 */
+            if (conn->actual_protocol_version >= S2N_TLS13 &&
+                conn->config->security_policy->certificate_signature_preferences->signature_schemes[i]->hash_alg == S2N_HASH_SHA1) {
```

Comment:
I'm not a fan of how specific this check is. Can we make it more general?

Unfortunately it looks like we can't necessarily use the supported version field of signature schemes (https://github.com/awslabs/s2n/blob/main/tls/s2n_signature_scheme.h#L29-L30), because that would ban PKCS1 signatures in TLS1.3 too, and PKCS1 signatures are still allowed on certificates. :(

Reply:
This is something that is explicitly called out in the RFC, so I assumed this check would be reasonable. But besides, there's not really another way we can allow SHA-1 legacy signatures in TLS1.2 and not in TLS1.3?

Reply:
That's fair. I guess anything we did now (like add a max_protocol_version to the hash algorithm itself) would be overkill for one restriction. We can always change this if future versions deprecate more algorithms.